### DISCUSSION TITLE
how to forward statusCode from axios request to nextjs

### DISCUSSION THREAD

**Discussion by Innei** [vercel/next.js#12337] [Help] [answered]
example:
```ts
Page.getInitialProps = async (ctx) => {
  const { data } = await Rest('Page', 'slug').get<PageRespDto>( // this is a axios instance
    ctx.query.page as string,
  )       
// in production
// if this request got 404, nextjs render nextjs Error page, 
//.  but the status code is 500, not 404 from axios request.
  return { data } as PageRespDto
}
```

**Comment by Innei** [comment]
and my _error.tsx

```tsx
import { useRouter } from 'next/router'

function _Error({ statusCode }) {
  const router = useRouter()
  return (
    <div className="error">
      <div className="error_wrap">
        <h1>{statusCode}</h1>
        <div className="desc">
          {statusCode === 404 ? (
            <h2>页面走丢了</h2>
          ) : (
            <h2>抱歉, 出了点小问题</h2>
          )}
        </div>
      </div>
      <p>去别处看看?</p>
      <button className={'btn red'} onClick={(_) => router.push('/')}>
        回到首页
      </button>
    </div>
  )
}

_Error.getInitialProps = ({ res, err }) => {
  const statusCode = res ? res.statusCode : err ? err.statusCode : 404
  return { statusCode }
}

export default _Error
```

**Comment by mspix** [comment, marked as answer]
I'm facing the same problem. How do we forward a statusCode from an axios request to next.js?
Does the error object need to be passed through axios .catch(error=> { ... }) to next.js? If yes, how?

**Reply by Innei** [reply]
Emm, this is my solution now. 

In axios utils, used axios response interceptors. and handle the response error like this.

```
// request.ts

service.interceptors.response.use(
  (response) => {
    const res = response.data

    return res
  },
  (error) => {
         // do anything...
    return Promise.reject({
      statusCode: error.response.status,
      data: error.response.data,
    })
  },
)
```

And in nextjs side, the `_error.tsx` file, do this.

```tsx
// import ...
const Error: FC = () => {
	// ... jsx
}

const getCode = (err, res) => {
  if (!err && !res) {
    return 500
  }
  if (res?.statusCode === 500 && err?.statusCode === 500) {
    return 500
  } else if (err && err.statusCode !== 500) {
    return err.statusCode
  } else if (res && res.statusCode !== 500) {
    return res.statusCode
  }
  return 500
}


Error.getInitialProps = ({ res, err }) => {
  const statusCode = getCode(err, res)

  return { statusCode } as { statusCode: number }
}
```

Why do this, why pick the error code which is not `500`.

In development mode, nextjs didn't do optimize pages, so the response error code will be raw code, but in production mode, the response code seem to be always `500`, but error code not. I don't know why. But I think this is worked.

**Reply by mspix** [reply]
Thank you Innei. Didn't know of axios interceptors before.

The interceptor is working for me, but the returned Promise.reject( ... ) doesn't trigger the Next.js Error Boundary, because it isn't called directly in the lifecycle of a component. A store is between the request function and the component. Somehow I have to separate the data fetch logic from the store or I get the error forwarding to Next.js Error Boundary to work.

Any way, thank you. It helped me moving forward.
